### DISCUSSION TITLE
I was creating a PHP calculator and ran into a bug

### DISCUSSION THREAD

**Discussion by ghost** [community/community#63588] [Programming Help] [answered]
The provided PHP code creates a calculator program that allows the user to perform basic arithmetic operations. Here's a detailed explanation of what the code does:

1\. The code defines a function called \`calc\` that takes a single argument \`$val\`. This function is responsible for handling the user input and performing the corresponding arithmetic operations.

1\. An array called \`$operators\` is defined, which contains the valid arithmetic operations: "add," "subtract," "multiply," "exit," "divide," and "remainder."

1\. The code enters a \`for\` loop that iterates through each element in the \`$operators\` array. It checks if the user input \`$val\` matches any of the operators.

1\. If the user input matches an operator, the code executes the corresponding block of code for that operator.

1\. If the operator is "exit," the code prompts the user to confirm if they want to exit the program. If the user confirms with "Y," the loop is broken, and the program terminates. If the user enters any other input, they are prompted to enter another type of math, and the \`calc\` function is called recursively with the new input.

1\. If the operator is "add," the code prompts the user to enter two numbers, \`$num1\` and \`$num2\`. The sum of the two numbers is then calculated and displayed. The user is prompted to enter another type of math, and the \`calc\` function is called recursively with the new input.

1\. If the operator is "subtract," the code prompts the user to enter two numbers, \`$num1\` and \`$num2\`. The difference between the two numbers is calculated and displayed. The user is prompted to enter another type of math, and the \`calc\` function is called recursively with the new input.

1\. If the operator is "multiply," the code prompts the user to enter two numbers, \`$num1\` and \`$num2\`. The product of the two numbers is calculated and displayed. The user is prompted to enter another type of math, and the \`calc\` function is called recursively with the new input.

1\. If the operator is "divide," the code prompts the user to enter two numbers, \`$num1\` and \`$num2\`. The code checks if either of the numbers is zero. If either of them is zero, an exception is thrown with the message "Division by zero!" Otherwise, the division of \`$num1\` by \`$num2\` is calculated and displayed. If an exception is caught, the error message is displayed. The user is prompted to enter another type of math, and the \`calc\` function is called recursively with the new input.

1\. If the operator is "remainder," the code prompts the user to enter two numbers, \`$num1\` and \`$num2\`. The code checks if \`$num1\` is divisible by \`$num2\`. If it is, the remainder of the division is calculated and displayed. If it is not divisible, a message indicating that is displayed. The user is prompted to enter another type of math, and the \`calc\` function is called recursively with the new input.

1\. If the user input does not match any of the operators, the code handles the error by displaying an appropriate message. If the input is empty, it prompts the user to enter a type of math. Otherwise, it displays an error message indicating that the input is unknown and provides a list of valid operators. The user is prompted to enter another type of math, and the \`calc\` function is called recursively with the new input.

1\. The code outside the \`calc\` function displays the available operators using a loop and prompts the user to enter a type of math they want to perform.

1\. The user's input is stored in the variable \`$typeOfMath\`, and the \`calc\` function is called with this input to begin the calculation process.

In summary, this PHP code creates a simple calculator program that allows the user to perform basic arithmetic operations. The program repeatedly prompts the user to enter a type of math, performs the requested operation, and displays the result.

```
<?php
function calc($val){
    $operators = ["add", "subtract", "multiply", "exit", "divide", "remainder"];
    for($i = 0; $i < count($operators); $i++){
        if($val == $operators[$i]){
            if($val == "exit"){
                echo "Are you sure you want to exit? Y/N > ";
                $confirmExit = readline();
                if($confirmExit == "Y"){
                    break;
                }else{
                    echo "Enter a type of math > ";
                    $typeOfMath = readline();
                    calc($typeOfMath);
                }
            }
            if($val == "add"){
                echo "Enter a number > ";
                $num1 = readline();
                echo "Enter another number > ";
                $num2 = readline();
                echo $num1+$num2 . "\n";
                echo "Enter another type of math > ";
                $val = readline();
                calc($val);
            }
            if($val == "subtract"){
                echo "Enter a number > ";
                $num1 = readline();
                echo "Enter another number > ";
                $num2 = readline();
                echo $num1-$num2 . "\n";
                echo "Enter another type of math > ";
                $val = readline();
                calc($val);
            }
            if($val == "multiply"){
                echo "Enter a number > ";
                $num1 = readline();
                echo "Enter another number > ";
                $num2 = readline();
                echo $num1*$num2 . "\n";
                echo "Enter another type of math > ";
                $val = readline();
                calc($val);
            }
            if($val == "divide"){
                try{
                    echo "Enter a number > ";
                    $num1 = readline();
                    echo "Enter another number > ";
                    $num2 = readline();
                    if($num1 || $num2 == 0){
                        throw new Exception("Division by zero!");
                    }else{
                        echo $num1/$num2;
                    }
                }catch(Exception $err){
                    echo $err;
                }
            }
            if($val == "remainder"){
                echo "Enter a number > ";
                $num1 = readline();
                echo "Enter another number > ";
                $num2 = readline();
                if($num1%$num2){
                    echo $num1%$num2 . "\n";
                }else{
                    echo "num1 isn't divisible by num2 \n";
                }
                echo "Enter another type of math > ";
                $val = readline();
                calc($val);
            }
        }else{
            if($val == ""){
                echo "The value cannot be empty! \n";
                echo "Enter a type of math > ";
                $typeOfMath = readline();
                calc($typeOfMath);
            }else{
                echo "The value " .$val ." is unkown! \n";
                echo "The existing values are: \n";
                for($j = 0; $j < count($operators); $j++){
                    echo $operators[$j] . "\n";
                }
                echo "Enter a type of math > ";
                $typeOfMath = readline();
                calc($typeOfMath);
            }
        }
    }
}
echo "The functions are \n ";
$operators = ["add", "subtract", "multiply", "divide", "remainder"];
for($i = 0; $i < count($operators); $i++){
    echo $operators[$i] ."\n";
}
echo "Enter a type of math you want to do > ";
$typeOfMath = readline();
calc($typeOfMath);
?>
```

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by AbdrrahimDahmani** [comment]
could you please provide me the source code repository so I can look clearly your code. And explain clearly the issue!

**Comment by denisbevilacqua** [comment]
It seems like you've implemented various arithmetic operations such as addition, subtraction, multiplication, division, and calculating the remainder. You've also included error handling for division by zero and provided a list of valid operators.

However, you mentioned that you ran into a bug. Could you please specify what the bug is or provide more details about the issue you're facing?

**Reply by ghost** [reply]
Whenever you enter an operator besides add it says that the operator doesn’t exist and gives you a list of operators

**Reply by denisbevilacqua** [reply, marked as answer]
In the section where you handle an unknown value, you should move the loop for displaying existing operators outside of the loop that checks for the value. Otherwise, you will display the list of existing operators multiple times if the user enters an unknown value multiple times. You can move the loop like this:

```php
if ($val == "") {
    echo "The value cannot be empty! \n";
} else {
    echo "The value " . $val . " is unknown! \n";
}
echo "The existing values are: \n";
for ($j = 0; $j < count($operators); $j++) {
    echo $operators[$j] . "\n";
}
echo "Enter a type of math > ";
$typeOfMath = readline();
calc($typeOfMath);
```

You can also improve the code by adding a Division by Zero Check and handling division by Zero. You can try it with this:

```php
<?php
function calc($val){
    $operators = ["add", "subtract", "multiply", "exit", "divide", "remainder"];
    for($i = 0; $i < count($operators); $i++){
        if($val == $operators[$i]){
            if($val == "exit"){
                echo "Are you sure you want to exit? Y/N > ";
                $confirmExit = readline();
                if($confirmExit == "Y"){
                    break;
                }else{
                    echo "Enter a type of math > ";
                    $typeOfMath = readline();
                    calc($typeOfMath);
                }
            }
            if($val == "add"){
                echo "Enter a number > ";
                $num1 = readline();
                echo "Enter another number > ";
                $num2 = readline();
                echo $num1 + $num2 . "\n";
                echo "Enter another type of math > ";
                $val = readline();
                calc($val);
            }
            if($val == "subtract"){
                echo "Enter a number > ";
                $num1 = readline();
                echo "Enter another number > ";
                $num2 = readline();
                echo $num1 - $num2 . "\n";
                echo "Enter another type of math > ";
                $val = readline();
                calc($val);
            }
            if($val == "multiply"){
                echo "Enter a number > ";
                $num1 = readline();
                echo "Enter another number > ";
                $num2 = readline();
                echo $num1 * $num2 . "\n";
                echo "Enter another type of math > ";
                $val = readline();
                calc($val);
            }
            if($val == "divide"){
                try{
                    echo "Enter a number > ";
                    $num1 = readline();
                    echo "Enter another number > ";
                    $num2 = readline();
                    if($num2 == 0){
                        throw new Exception("Division by zero!");
                    }else{
                        echo $num1 / $num2;
                    }
                }catch(Exception $err){
                    echo $err;
                }
            }
            if($val == "remainder"){
                echo "Enter a number > ";
                $num1 = readline();
                echo "Enter another number > ";
                $num2 = readline();
                if($num2 != 0){
                    echo $num1 % $num2 . "\n";
                }else{
                    echo "num1 isn't divisible by num2 \n";
                }
                echo "Enter another type of math > ";
                $val = readline();
                calc($val);
            }
        }
    }

    // Handle unknown or empty value
    if($val == ""){
        echo "The value cannot be empty! \n";
    }else{
        echo "The value " .$val ." is unknown! \n";
        echo "The existing values are: \n";
        for($j = 0; $j < count($operators); $j++){
            echo $operators[$j] . "\n";
        }
    }
    echo "Enter a type of math > ";
    $typeOfMath = readline();
    calc($typeOfMath);
}

echo "The functions are \n ";
$operators = ["add", "subtract", "multiply", "divide", "remainder", "exit"];
for($i = 0; $i < count($operators); $i++){
    echo $operators[$i] ."\n";
}
echo "Enter a type of math you want to do > ";
$typeOfMath = readline();
calc($typeOfMath);
?>
```
